Make genindex support all kinds of letters, not only Latin ones

mitya57 opened the pull request:
Without this i.e. all names starting with a Cyrillic letter are moved into a Symbols section rather than being split into sections by letter. [Example page](http://docs.istina.msu.ru/genindex.html).

Before | After
------ | ------
![genindex](https://cloud.githubusercontent.com/assets/1381584/23342720/d918763a-fc78-11e6-84a5-3aa0541bb241.png) | ![genindex-new](https://cloud.githubusercontent.com/assets/1381584/23342721/d91c2e2e-fc78-11e6-89a3-cdac03f65241.png)

mitya57 committed:
Make genindex support all kinds of letters, not only Latin ones

Without this i.e. all names starting with a Cyrillic letter are moved
into a Symbols section rather than being split into sections by letter.

tk0miya commented on sphinx/environment/managers/indexentries.py:
```diff
@@ -114,13 +113,13 @@ def add_entry(word, subword, main, link=True, dic=new, key=None):
         # sort the index entries; put all symbols at the front, even those
         # following the letters in ASCII, this is where the chr(127) comes from
-        def keyfunc(entry, lcletters=string.ascii_lowercase + '_'):
+        def keyfunc(entry, lcletters=None):
```
Is `lcletters` still needed?

tk0miya commented on sphinx/environment/managers/indexentries.py:
```diff
@@ -158,14 +157,14 @@ def keyfunc(entry, lcletters=string.ascii_lowercase + '_'):
 
         # group the entries by letter
-        def keyfunc2(item, letters=string.ascii_uppercase + '_'):
+        def keyfunc2(item, letters=None):
```
Is `letters` also needed?

tk0miya reviewed (changes requested):
LGTM with nits.
BTW could you add testcase to `tests/test_environment_indexentries.py` ?

tk0miya commented:
@shimizukawa This might be related with the key for glossaries. So please give us comment if you have.

mitya57 committed:
Make genindex support RTL languages

If a string starts with an RTL mark, simply skip it and use the following
character as the first letter.

mitya57 committed:
Add tests for the previous two changes

mitya57 commented:
@tk0miya Updated! It should also support RTL languages now: I added a test that checks that the first letter of ‏תירבע‎ is ת.

mitya57 commented:
Note for future: ideally uppercase should be language-aware. I.e. uppercase i is I in most languages, but it is İ in Turkish.

Also the canonical decomposition is not always the best way to go: for example, S and Š are usually considered distinct letters (and popular) in Czech, Lithuanian and Latvian, and should not be joined together. In Russian, И and Й should be also considered distinct letters (unlike Е and Ё, which should usually be grouped together).

But this would not be easy to fix.

shimizukawa commented:
> This might be related with the key for glossaries. So please give us comment if you have.

I checked and LGTM. 
Glossary key `glossary term : glossary key` feature is not affected by this change.

tk0miya commented:
LGTM!
I will merge this later (because adding CHANGES is needed after merging, but I don't have console now).

tk0miya commented:
Merged, Thanks!
